lib/CL/pocl_util.c

```diff
@@ -1117,95 +923,136 @@ pocl_create_command_full (_cl_command_node **cmd,
-        continue;
+        /* If both a content buffer and its associated size buffer are
+         * explicitly listed in buffers, the size buffer was already migrated
+         * above. */
+        if (already_migrated[i])
+          {
+            ++i;
+            continue;
+          }
 
-      if (buffers[i]->size_buffer != NULL)
-        {
-          /* BLOCK until size buffer has been imported to host mem! No event
-           * exists if host import is not needed. */
-          if (size_events[i] != NULL)
-            {
-              cl_device_id d = size_events[i]->queue->device;
-              POname (clWaitForEvents) (1, &size_events[i]);
-              if (buffers[i]->size_buffer->mem_host_ptr != NULL)
-                {
-                  migration_size
-                      = *(uint64_t *)buffers[i]->size_buffer->mem_host_ptr;
-                }
-              pocl_release_mem_host_ptr (buffers[i]->size_buffer);
-              POname (clReleaseEvent) (size_events[i]);
-              size_events[i] = NULL;
-            }
-        }
+        if (mi->buffer->size_buffer != NULL)
+          {
+            /* BLOCK until size buffer has been imported to host mem! No event
+             * exists if host import is not needed. */
+            if (size_events[i] != NULL)
+              {
+                cl_device_id d = size_events[i]->queue->device;
+                POname (clWaitForEvents) (1, &size_events[i]);
+                if (mi->buffer->size_buffer->mem_host_ptr != NULL)
+                  {
+                    migration_size
+                      = *(uint64_t *)mi->buffer->size_buffer->mem_host_ptr;
+                  }
+                pocl_release_mem_host_ptr (mi->buffer->size_buffer);
+                POname (clReleaseEvent) (size_events[i]);
+                size_events[i] = NULL;
+              }
+          }
 
-      pocl_create_migration_commands (
-          dev, NULL, final_event, buffers[i],
-          &buffers[i]->device_ptrs[dev->global_mem_id], readonly_flags[i],
+        pocl_create_migration_commands (
+          dev, NULL, final_event, mi->buffer,
+          &mi->buffer->device_ptrs[dev->global_mem_id], mi->read_only,
           command_type, mig_flags, migration_size);
-    }
 
-  return err;
+        /* Hold the last updater events of the parent buffers so we can refer
+           to the event in the possible patch sub-buffers. These will point to
+           the potential new migration commands.  */
+        if (mi->buffer->parent != NULL)
+          POname (clRetainEvent) (mi->buffer->last_updater);
+        ++i;
+      }
+
+    LL_FOREACH (buffer_usage, mi)
+      {
+        /* The last events of the parent buffers can be now released as the
+           sub-buffer references to the events should hold them alive. */
+        if (mi->buffer->parent == NULL && mi->buffer->sub_buffers != NULL)
+          {
+            int new_ref;
+            POCL_RELEASE_OBJECT_UNLOCKED (mi->buffer->parent->last_updater,
```

Comment:
This only decrements the refcount and does NOT actually release anything. The next line seems to assume it does, though?